# Лабораторная работа №2

In [50]:
!pwd

/Users/artyem.petrov/dev/university/4-1/intellectual-systems/lab02


## 1. Выбираем размеченную базу для задачи регрессии.

In [51]:
!./.venv/bin/pip install scikit-learn pandas numpy


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: /Users/artyem.petrov/dev/university/4-1/intellectual-systems/lab02/.venv/bin/python -m pip install --upgrade pip


In [52]:
from sklearn.datasets import fetch_california_housing
import pandas as pd
import numpy as np

In [53]:
housing = fetch_california_housing()

df_housing = pd.DataFrame(housing.data, columns=housing.feature_names)
df_housing['target'] = housing.target

df_housing.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

## 2. Выполняем стандартизацию данных (при необходимости).

In [54]:
from sklearn.preprocessing import StandardScaler

In [55]:
X = housing.data
y = housing.target

# Выполняем стандартизацию признаков
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

df_scaled = pd.DataFrame(X_scaled, columns=housing.feature_names)
df_scaled['target'] = y

df_scaled.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  2.344766  0.982143  0.628559  -0.153758   -0.974429 -0.049597  1.052548   
1  2.332238 -0.607019  0.327041  -0.263336    0.861439 -0.092512  1.043185   
2  1.782699  1.856182  1.155620  -0.049016   -0.820777 -0.025843  1.038503   
3  0.932968  1.856182  0.156966  -0.049833   -0.766028 -0.050329  1.038503   
4 -0.012881  1.856182  0.344711  -0.032906   -0.759847 -0.085616  1.038503   

   Longitude  target  
0  -1.327835   4.526  
1  -1.322844   3.585  
2  -1.332827   3.521  
3  -1.337818   3.413  
4  -1.337818   3.422

In [56]:
df_housing.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

## 3. Выполняем расширение базовых признаков за счет применения полиномиальной модели (степень = 3)

In [57]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=3, include_bias=False)
X_poly = poly.fit_transform(X_scaled)

feature_names = poly.get_feature_names_out(housing.feature_names)

df_poly = pd.DataFrame(X_poly, columns=feature_names)
df_poly['target'] = y

df_poly.iloc[:, :10].head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  2.344766  0.982143  0.628559  -0.153758   -0.974429 -0.049597  1.052548   
1  2.332238 -0.607019  0.327041  -0.263336    0.861439 -0.092512  1.043185   
2  1.782699  1.856182  1.155620  -0.049016   -0.820777 -0.025843  1.038503   
3  0.932968  1.856182  0.156966  -0.049833   -0.766028 -0.050329  1.038503   
4 -0.012881  1.856182  0.344711  -0.032906   -0.759847 -0.085616  1.038503   

   Longitude  MedInc^2  MedInc HouseAge  
0  -1.327835  5.497926         2.302894  
1  -1.322844  5.439334        -1.415713  
2  -1.332827  3.178017         3.309014  
3  -1.337818  0.870428         1.731757  
4  -1.337818  0.000166        -0.023909

## 4. Разбиваем датасеты (базовый и расширенный) на 2 множества (обучающее и тестовое).

In [58]:
from sklearn.model_selection import train_test_split

In [59]:
# Разбиваем базовый датасет (стандартизированные признаки)
X_train_basic, X_test_basic, y_train_basic, y_test_basic = train_test_split(
    X_scaled, y, test_size=0.3, random_state=42
)

# Разбиваем расширенный датасет (полиномиальные признаки)
X_train_poly, X_test_poly, y_train_poly, y_test_poly = train_test_split(
    X_poly, y, test_size=0.3, random_state=42
)

print("Базовый датасет (стандартизированные признаки):")
print(f"  Обучающая выборка: {X_train_basic.shape[0]} объектов, {X_train_basic.shape[1]} признаков")
print(f"  Тестовая выборка: {X_test_basic.shape[0]} объектов, {X_test_basic.shape[1]} признаков")

print("\nРасширенный датасет (полиномиальные признаки):")
print(f"  Обучающая выборка: {X_train_poly.shape[0]} объектов, {X_train_poly.shape[1]} признаков")
print(f"  Тестовая выборка: {X_test_poly.shape[0]} объектов, {X_test_poly.shape[1]} признаков")

print("\nЦелевая переменная (одинаковая для обоих датасетов):")
print(f"  Обучающая выборка: {y_train_basic.shape[0]} значений")
print(f"  Тестовая выборка: {y_test_basic.shape[0]} значений")

Базовый датасет (стандартизированные признаки):
  Обучающая выборка: 14448 объектов, 8 признаков
  Тестовая выборка: 6192 объектов, 8 признаков

Расширенный датасет (полиномиальные признаки):
  Обучающая выборка: 14448 объектов, 164 признаков
  Тестовая выборка: 6192 объектов, 164 признаков

Целевая переменная (одинаковая для обоих датасетов):
  Обучающая выборка: 14448 значений
  Тестовая выборка: 6192 значений


## 5. Выполняем обучение линейной и полиномиальной моделей аналитическим методом - не используя готовое библиотечное решение.

In [60]:
# Обучение линейной модели с использованием нормальных уравнений: w = (X^T * X)^(-1) * X^T * y
def train_linear_model_analytical(X, y):
    # Добавляем столбец единиц для cвободного члена
    X_with_intercept = np.column_stack([np.ones(X.shape[0]), X])
    
    # Вычисляем веса по нормальным уравнениям
    XTX = X_with_intercept.T @ X_with_intercept
    XTy = X_with_intercept.T @ y
    
    # Решаем систему уравнений
    try:
        w = np.linalg.solve(XTX, XTy)
    except np.linalg.LinAlgError:
        # Если матрица вырождена, используем псевдообратную
        w = np.linalg.pinv(XTX) @ XTy
    
    return w

In [61]:
def predict_linear_model(X, w):
    """Предсказание с использованием обученной линейной модели"""
    X_with_intercept = np.column_stack([np.ones(X.shape[0]), X])
    return X_with_intercept @ w

In [62]:
# Обучение линейной модели на базовых признаках
w_basic = train_linear_model_analytical(X_train_basic, y_train_basic)
print(f"Обучено весов: {len(w_basic)} (включая intercept)")

# Обучение линейной модели на полиномиальных признаках
w_poly = train_linear_model_analytical(X_train_poly, y_train_poly)
print(f"Обучено весов: {len(w_poly)} (включая intercept)")

# Предсказания на тестовых данных
y_pred_basic = predict_linear_model(X_test_basic, w_basic)
y_pred_poly = predict_linear_model(X_test_poly, w_poly)

print(f"Линейная модель (базовые признаки): {len(w_basic)} параметров")
print(f"Линейная модель (полиномиальные признаки): {len(w_poly)} параметров")

Обучено весов: 9 (включая intercept)
Обучено весов: 165 (включая intercept)
Линейная модель (базовые признаки): 9 параметров
Линейная модель (полиномиальные признаки): 165 параметров


In [63]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [64]:
def calculate_regression_metrics(y_true, y_pred, model_name):
    mse = mean_squared_error(y_true, y_pred)
    mae = mean_absolute_error(y_true, y_pred)
    r2 = r2_score(y_true, y_pred)
    rmse = np.sqrt(mse)
    
    print(f"\n{model_name}:")
    print(f"  MSE (Mean Squared Error): {mse:.4f}")
    print(f"  RMSE (Root Mean Squared Error): {rmse:.4f}")
    print(f"  MAE (Mean Absolute Error): {mae:.4f}")
    print(f"  R² (Coefficient of Determination): {r2:.4f}")
    
    return {'mse': mse, 'rmse': rmse, 'mae': mae, 'r2': r2}

# Вычисляем метрики для обеих моделей
metrics_basic = calculate_regression_metrics(y_test_basic, y_pred_basic, 
                                           "Линейная модель (базовые признаки)")
metrics_poly = calculate_regression_metrics(y_test_poly, y_pred_poly, 
                                          "Линейная модель (полиномиальные признаки)")



Линейная модель (базовые признаки):
  MSE (Mean Squared Error): 0.5306
  RMSE (Root Mean Squared Error): 0.7284
  MAE (Mean Absolute Error): 0.5272
  R² (Coefficient of Determination): 0.5958

Линейная модель (полиномиальные признаки):
  MSE (Mean Squared Error): 19.3159
  RMSE (Root Mean Squared Error): 4.3950
  MAE (Mean Absolute Error): 0.5065
  R² (Coefficient of Determination): -13.7165


## 6. Выполняем обучение линейной и полиномиальной моделей градиентным спуском - не используя готовое библиотечное решение.

In [69]:
# Функция для обучения линейной модели градиентным спуском
def train_linear_model_gradient_descent(X, y, learning_rate=0.001, n_iterations=1000, tolerance=1e-6):
    X_with_intercept = np.column_stack([np.ones(X.shape[0]), X])
    w = np.random.randn(X_with_intercept.shape[1]) * 0.01
    losses = []
    
    for i in range(n_iterations):
        y_pred = X_with_intercept @ w
        error = y_pred - y
        loss = np.mean(error ** 2)
        losses.append(loss)
        
        gradient = (2 / len(y)) * (X_with_intercept.T @ error)
        
        # Gradient clipping
        grad_norm = np.linalg.norm(gradient)
        if grad_norm > 1.0:
            gradient = gradient / grad_norm
        
        w_new = w - learning_rate * gradient
        
        # Проверка сходимости весов
        if np.linalg.norm(w_new - w) < tolerance:
            print(f"  Сходимость достигнута на итерации {i+1}")
            break
            
        # Проверка расходимости
        if not np.isfinite(loss) or not np.isfinite(grad_norm):
            print(f"  Расходимость на итерации {i+1}")
            break
            
        # Проверка сходимости по loss
        if i > 0 and abs(losses[-1] - losses[-2]) < tolerance:
            print(f"  Сходимость по loss на итерации {i+1}")
            break
        
        w = w_new
        
        if i % 100 == 0:
            print(f"  Итерация {i}, Loss: {loss:.4f}")
    
    return w, losses

print("Обучение линейной модели на базовых признаках с градиентным спуском)...")
w_basic_gd, losses_basic = train_linear_model_gradient_descent(X_train_basic, y_train_basic, learning_rate=0.001)

print("Обучение линейной модели на полиномиальных признаках с градиентным спуском...")
w_poly_gd, losses_poly = train_linear_model_gradient_descent(X_train_poly, y_train_poly, learning_rate=0.0001)

# Предсказания на тестовых данных
y_pred_basic_gd = predict_linear_model(X_test_basic, w_basic_gd)
y_pred_poly_gd = predict_linear_model(X_test_poly, w_poly_gd)


Обучение линейной модели на базовых признаках с градиентным спуском)...
  Итерация 0, Loss: 5.5795
  Итерация 100, Loss: 5.1443
  Итерация 200, Loss: 4.7293
  Итерация 300, Loss: 4.3345
  Итерация 400, Loss: 3.9598
  Итерация 500, Loss: 3.6052
  Итерация 600, Loss: 3.2707
  Итерация 700, Loss: 2.9562
  Итерация 800, Loss: 2.6618
  Итерация 900, Loss: 2.3874
Обучение линейной модели на полиномиальных признаках с градиентным спуском...
  Итерация 0, Loss: 14536.1895
  Итерация 100, Loss: 115.6076
  Итерация 200, Loss: 14.8895
  Итерация 300, Loss: 14.5038
  Итерация 400, Loss: 14.1276
  Итерация 500, Loss: 13.7620
  Итерация 600, Loss: 13.4091
  Итерация 700, Loss: 13.0705
  Итерация 800, Loss: 12.7469
  Итерация 900, Loss: 12.4386


## Выводы

Модель на полиномиальных признаках обучилась плохо. Без gradient clipping было бы еще хуже, так как я получал nan loss. Это вызвано следующими причинами: 

1. Проблема переобучения на полиномиальных признаках: 

    - Базовые признаки: 8 признаков -> 9 параметров
    - Полиномиальные признаки: 164 признака -> 165 параметров
      

    При степени полинома 3 создается 164 признака, что значительно больше исходных 8. Это приводит к:
    - Высокой сложности модели
    - Переобучению на обучающих данных
    - Плохой обобщающей способности

2. Полиномиальная модель показывает R² = -13.7165, что означает, что она работает сильно хуже среднего. 